operator/pkg/nodepools/pools.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package nodepools exists currently because it is required to be a separate due to import conflicts.
+package nodepools
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+
+	vectorizedv1alpha1 "github.com/redpanda-data/redpanda-operator/operator/api/vectorized/v1alpha1"
+	"github.com/redpanda-data/redpanda-operator/operator/pkg/labels"
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+func GetNodePools(ctx context.Context, cluster *vectorizedv1alpha1.Cluster, k8sClient client.Reader) ([]*vectorizedv1alpha1.NodePoolSpecWithDeleted, error) {
+	var nodePoolsWithDeleted []*vectorizedv1alpha1.NodePoolSpecWithDeleted
+
+	nps := cluster.GetNodePoolsFromSpec()
+	for i := range nps {
+		np := nps[i]
+
+		nodePoolsWithDeleted = append(nodePoolsWithDeleted, &vectorizedv1alpha1.NodePoolSpecWithDeleted{
+			NodePoolSpec: np,
+		})
+	}
+
+	// Also add "virtual NodePools" based on StatefulSets found.
```

Comment:
This comment is out of place. Either place it in the function comment or directly above the relevant code.

Reply:
Is it out of place? It is right above the statefulset lookup,and that is for sure part of this creation of the "virtual NodePools".

Reply:
AH. Wow, I really misread this function the first time around. I think what's really throwing me for a loop (aside from using side by side view in github) is the inconsistency of breaking loops within this loop and the number at that.

Could we pick a single method of iterating  and short circuiting be it labeled breaks, `slices.Contains`, or a helper `findFunc[T any]([]T, func(T) bool) (*T, bool)` function?